File: prog_java/network_plann/NetworkPlann.java
package network_plann;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.Set;


public class NetworkPlann {
    //Вычисление предшественников
    private ArrayList<ArrayList<Integer> > calculate_previous(int [][] matrix){
        ArrayList<ArrayList<Integer> > previous = new ArrayList<ArrayList<Integer>>(matrix.length);
        for(int i = 0; i< matrix.length; i++)
            previous.add(new ArrayList<Integer>());

        for(int i = 0; i < matrix.length; i++)
            for(int j = 0; j < matrix.length; j++)
                if (matrix[i][j] > 0)
                    previous.get(j).add(i);
        return previous;
    };

    //Вычисление ранних сроков свершения событий
    private ArrayList<Integer> calculate_Tp(int [][] matrix, ArrayList<ArrayList<Integer> > prev){
        ArrayList<Integer> Tp = new ArrayList<Integer>();
        for(int i = 0; i < matrix.length; i++)
            Tp.add(0);

        for(int i = 0; i < Tp.size(); i++)
        {
            ArrayList<Integer> previous = prev.get(i);
            ArrayList<Integer> temp = new ArrayList<Integer>();
            for (Integer j : previous) 
                if(previous.size() == 1 && j == 0)
                    Tp.set(i, matrix[j][i]);
                else
                    temp.add(matrix[j][i] + Tp.get(j));
            if(temp.size() != 0)
                Tp.set(i, Collections.max(temp));
        }
        return Tp;
    };

    //Вычисление поздних сроков свершения событий
    private ArrayList<Integer> calculate_Tr(int [][] matrix, ArrayList<Integer> Tp){
        ArrayList<Integer> Tr = new ArrayList<Integer>();
        for(int i = 0; i < matrix.length; i++)
            Tr.add(0);

        Tr.set(Tp.size() - 1, Tp.get(Tp.size() - 1));
        for(int i = Tp.size() - 2; i >= 0; i--){
            ArrayList<Integer> temp = new ArrayList<Integer>();
            for(int j = 0; j < Tr.size(); j++)
                if (matrix[i][j] > 0)
                    temp.add(Tr.get(j) - matrix[i][j]);
            Tr.set(i, Collections.min(temp));
        }
        return Tr;
    };

    //Вычисление критического пути
    private ArrayList<Integer> calculate_critical_path(ArrayList<Integer> Tp, ArrayList<Integer> Tr){
        ArrayList<Integer> critical_path = new ArrayList<Integer>();
        for(int i = 0; i < Tr.size(); i++)
            if (Tr.get(i) == Tp.get(i))
                critical_path.add(i);
        return critical_path;
    };

    //Проверка размера матрицы (размер должен быть n x n)
    private Boolean check_matrix(int [][] matrix){
        for(int i = 0; i < matrix.length; i++)
            if(matrix.length == matrix[i].length)
                continue;
            else 
                return false;
        return true;
    };

    //Вычисление всех резервов времени работы
    private ArrayList<HashMap<String, Integer>> calculate_all_reserve(int [][] matrix, ArrayList<Integer> Tp, ArrayList<Integer> Tr){
        ArrayList<HashMap<String, Integer>> reserve = new ArrayList<HashMap<String, Integer>>();
        HashMap<String, Integer> Rfull = new HashMap<String, Integer>();
        HashMap<String, Integer> Rguaranteed = new HashMap<String, Integer>();
        HashMap<String, Integer> Rfree = new HashMap<String, Integer>();
        HashMap<String, Integer> Rindependent = new HashMap<String, Integer>();
        for(int i = 0; i < matrix.length; i++){
            for(int j = 0; j < matrix.length; j++)
                if(matrix[i][j] > 0)
                {
                    Rfull.put(Integer.toString(i) + "-" + Integer.toString(j),  Tr.get(j) - Tp.get(i) - matrix[i][j]);
                    Rguaranteed.put(Integer.toString(i) + "-" + Integer.toString(j), Tr.get(j) - Tr.get(i) - matrix[i][j]);
                    Rfree.put(Integer.toString(i) + "-" + Integer.toString(j), Tp.get(j) - Tp.get(i) - matrix[i][j]);
                    Rindependent.put(Integer.toString(i) + "-" + Integer.toString(j), Tp.get(j) - Tr.get(i) - matrix[i][j]);  
                }
        }
        reserve.add(Rfull);
        reserve.add(Rguaranteed);
        reserve.add(Rfree);
        reserve.add(Rindependent);
        return reserve;
    }

    //Вывод ранних и поздних сроков свершения событий
    private void print_Tp_Tr(ArrayList<Integer> Tp, ArrayList<Integer> Tr){
        System.out.println("_____________Сроки свершения событий______________");
        System.out.printf("%-13s  ", "Событие");
        System.out.printf("%-13s  ", "Ранние");
        System.out.printf("%-13s  ", "Поздние");
        System.out.printf("%-13s  ", "Резерв");
        System.out.println(" ");
        for(int i = 0; i < Tp.size(); i++){
            System.out.printf("%-15s  ", i);
            System.out.printf("%-14s  ", Tp.get(i));
            System.out.printf("%-13s  ", Tr.get(i));
            System.out.printf("%-5s  ", Tr.get(i) - Tp.get(i));
            System.out.println(" ");
        }
        System.out.println(" ");
    }

    //Вывод резервов работы
    private void print_reserve(ArrayList<HashMap<String, Integer>> reserve){
        System.out.println("___________________Резервы работ_____________________");
        System.out.printf("%-5s  ", "Связь");
        System.out.printf("%-5s  ", "Полный");
        System.out.printf("%-5s  ", "Гарантированный");
        System.out.printf("%-5s  ", "Свободный");
        System.out.printf("%-5s  ", "Независимый");
        System.out.println(" ");
        Set<String> keys = reserve.get(0).keySet();

        for (String key : keys) {
            System.out.printf("%-8s  ", key);
            System.out.printf("%-10s  ", reserve.get(0).get(key));
            System.out.printf("%-12s  ", reserve.get(1).get(key));
            System.out.printf("%-10s  ", reserve.get(2).get(key));
            System.out.printf("%-5s  ", reserve.get(3).get(key));
            System.out.println(" ");
        }
        System.out.println(" ");
    }

    //Вывод маршрута критического пути
    private void print_critical_path(ArrayList<Integer> critical_path){
        System.out.print("Критический путь: ");
        for(int i = 0; i < critical_path.size(); i++){
            if(i < critical_path.size() - 1)
                System.out.print(" " + critical_path.get(i) + " -> ");
            else
            System.out.print(" " + critical_path.get(i));
        }
        System.out.println(" ");
    }
    
    //Вычисление параметров сетевой модели
    public void calculate_parameters(int [][] matrix){
        if (check_matrix(matrix)){
            ArrayList<ArrayList<Integer> > previous = calculate_previous(matrix);

            ArrayList<Integer> Tp = calculate_Tp(matrix, previous);
            ArrayList<Integer> Tr = calculate_Tr(matrix, Tp);
            print_Tp_Tr(Tp, Tr);
    
            ArrayList<HashMap<String, Integer>> reserve = calculate_all_reserve(matrix, Tp, Tr);
            print_reserve(reserve);

            ArrayList<Integer> critical_path = calculate_critical_path(Tp, Tr);
            print_critical_path(critical_path);
        }
        else
            System.out.println("Неверный размер матрицы!");
    }

}

File: prog_python/network_plann.py
#Вычисление предшественников
def calculate_previous(matrix: list[list]) -> list:
    previous = [[] for i in range(len(matrix))]
    for i, n_i in enumerate(matrix):
        for j, n_j in enumerate(n_i): 
            if type(n_j) == int:
                if n_j > 0:
                    previous[j].append(i)
            else:
                previous[j].append(i)
    return previous

#Вычисление ранних сроков свершения событий
def calculate_Tp(matrix: list[list], prev: list[list]) -> list:
    Tp = [0 for i in range(len(matrix))] # Установка размерности и заполнение нулями вектора ранних сроков
    for i in range(len(matrix)):
        previous = prev[i]
        temp = list()
        for j in previous:
            if len(previous) == 1 and j == 0:
                Tp[i] = matrix[j][i]
            else:
                temp.append(matrix[j][i] + Tp[j])
        print(temp)
        if temp:
            Tp[i] = max(temp)
    return Tp

#Вычисление поздних сроков свершения событий
def calculate_Tr(matrix: list[list], Tp: list) -> list:
    Tr = [0 for i in range(len(matrix))]  # Установка размерности и заполнение нулями вектора поздних сроков
    Tr[len(Tp) - 1] = Tp[len(Tp) - 1]
    for i in range(len(Tr) - 2, -1, -1):
        temp = list()
        for j in range(len(Tr)):
            if matrix[i][j] > 0:
                temp.append(Tr[j] - matrix[i][j])
        Tr[i] = min(temp)
    return Tr            

#Вычисление критического пути
def calculate_critical_path(Tr: list, Tp: list) -> list:
    critical_path = list()
    for i in range(len(Tr)):
        if Tr[i] == Tp[i]:
            critical_path.append(i)
    return critical_path
    
#Проверка размера матрицы (размер должен быть n x n)
def check_matrix(matrix: list[list]) -> bool:
    for i in range(len(matrix)):
        if len(matrix) == len(matrix[i]):
            continue
        else:
            return False
    return True

#Вычисление всех резервов времени работы
def calculate_all_reserve(Tp: list, Tr: list, matrix: list[list]) -> tuple[dict, dict, dict, dict]:
    Rfull = dict() # Полный резерв времени работы
    Rguaranteed = dict() # Гарантированный резерв времени работы    
    Rfree = dict() # Свободный резерв времени работы
    Rindependent = dict() # Независимый резерв времени работы
    for i in range(len(matrix)):
        for j in range(len(matrix)):
            if matrix[i][j] > 0:
                Rfull[f'{i}' + '-' + f'{j}'] = Tr[j] - Tp[i] - matrix[i][j]
                Rguaranteed[f'{i}' + '-' + f'{j}'] = Tr[j] - Tr[i] - matrix[i][j]
                Rfree[f'{i}' + '-' + f'{j}'] = Tp[j] - Tp[i] - matrix[i][j]
                Rindependent[f'{i}' + '-' + f'{j}'] = Tp[j] - Tr[i] - matrix[i][j]
    return Rfull, Rguaranteed, Rfree, Rindependent

#Вывод ранних и поздних сроков свершения событий
def print_Tp_Tr(Tp: list, Tr: list) -> None:
    print('__________Сроки свершения событий___________')
    column_name = ['Событие', 'Ранние', 'Поздние', 'Резерв']
    print(f'{column_name[0]:^10}' + "|" + f'{column_name[1]:^10}' + "|" + f'{column_name[2]:^10}' + "|" + f'{column_name[3]:^10}' + "|")
    for i in range(len(Tp)):
        print(f'{i:^10}' + "|" + F'{Tp[i]:^10}' + "|" + F'{Tr[i]:^10}' + "|"+ F'{Tr[i] - Tp[i]:^10}' + "|")
    print("\n")

#Вывод маршрута критического пути
def print_critical_path(critical_path: list) -> None:
    print('Критический путь:')
    for i in range(len(critical_path)):
        if(i != len(critical_path) - 1):
            print(critical_path[i], " -> ", end = '')
        else:
            print(critical_path[i], end = '')
    print("\n\n")

#Вывод резервов работы
def print_reserve(Rfull: dict, Rguaranteed: dict, Rfree: dict, Rindependent: dict) -> None:
    print('___________________________Резервы работ___________________________')
    column_name = ['Связь', 'Полный', 'Гарантированный', 'Свободный', 'Независимый']

    print(f'{column_name[0]:^10}' + "|" + f'{column_name[1]:^10}' + "|" + f'{column_name[2]:^17}' + "|" + f'{column_name[3]:^12}' + "|" + f'{column_name[4]:^14}' + "|")
    for key in Rfull:
        print(f'{key:^10}' + "|" + f'{Rfull[key]:^10}' + "|" + f'{Rguaranteed[key]:^17}' + "|"  + f'{Rfree[key]:^12}' + "|"  + f'{ Rindependent[key]:^14}' + "|")
    print("\n")

#Вычисление параметров сетевой модели
def calculate_parameters(matrix: list[list]) -> None:
    if check_matrix(matrix):
        prev = calculate_previous(matrix)
        Tp = calculate_Tp(matrix, prev)
        Tr = calculate_Tr(matrix, Tp)
        print_Tp_Tr(Tp, Tr)

        Rfull, Rguaranteed, Rfree, Rindependent = calculate_all_reserve(Tp, Tr, matrix)
        print_reserve(Rfull, Rguaranteed, Rfree, Rindependent)

        critical_path = calculate_critical_path(Tr, Tp)
        print_critical_path(critical_path)
    else:
        print('Неверный размер матрицы!')